thirdweb/modules/marketplace.py

```diff
@@ -0,0 +1,403 @@
+            opts["value"] = amount
+        else:
+            currency_module = ERC20(self.client, self.address)
+            allowance = currency_module.allowance.call(
+                self.get_signer_address(), self.address)
+            if allowance < amount:
+                tx = currency_module.increase_allowance.build_transaction(
+                    self.address, amount - allowance, opts)
+                self.execute_tx(tx)
+        return opts
+
+    def make_auction_listing_bid(self, bid: NewBid):
+        listing = self.validate_auction_listing(bid.listing_id)
+
+        bid_buffer = self.get_bid_buffer_bps()
+        winning_bid = self.get_winning_bid(bid.listing_id)
+        if winning_bid is not None:
+            is_winning_bidder = self.is_winning_bid(
+                winning_bid.pricePerToken, bid.price_per_token, bid_buffer)
+            if not is_winning_bidder:
+                raise Exception("The bid price is not within the bid buffer")
+        self.set_allowance(bid.price_per_token * listing.quantity,
+                           listing.currency_contract_address, self.get_transact_opts())
+        tx = self.__abi_module.offer.build_transaction(
+            bid.listing_id, listing.quantity, listing.currency_contract_address, bid.price_per_token)
+        self.execute_tx(tx)
+
+    def is_winning_bid(self, winning_price: int, new_bid_price: int, bid_buffer: int):
+        buffer = (new_bid_price * self.MAX_BPS) / winning_price
+        return (buffer > bid_buffer)
+
+    def get_winning_bid(self, listing_id: int):
+        """
+        Get the winning bid for a listing.
+
+        :param listing_id: The listing id.
+        :return: The winning bid.
+        """
+        bid = self.__abi_module.winning_bid.call(listing_id)
+
+        if bid[1] == ZeroAddress:
+            return None
+
+        if bid is None:
+            return None
+        return self.map_offer(listing_id, bid)
+
+    def get_direct_listing(self, listing_id: int):
+        listing = self.__abi_module.listings.call(listing_id)
+        if listing[0] != listing_id:
+            raise AssetNotFoundException(listing_id)
+        if listing[11] != 1:
+            raise WrongListingTypeException(
+                listing_id, "Auction", "Direct")
+
+        return self.map_direct_listing(listing)
+
+    def map_direct_listing(self, listing):
+        nftcontract = NftModule(listing[3], self.get_client())
+        return DirectListing(listing[0], assetContractAddress=listing[2], tokenId=listing[3], asset=nftcontract.get(listing[3]), startTimeInSeconds=listing[4], secondsUntilEnd=listing[5], quantity=listing[6], currencyContractAddress=listing[7], buyoutCurrencyValuePerToken=CurrencyModule(listing[7], self.get_client).get_value(listing[8]), buyoutPrice=listing[9], sellerAddress=listing[1], listing_type=[11])
+        
+    def _handle_token_approval(self, asset: str, token_id: int, from_address: str):
```

Comment:
this and `map_direct_listing` should be private I believe? same for a bunch of the internal functions below